external/fv3fit/fv3fit/pytorch/recurrent/train_fmr.py

```diff
@@ -0,0 +1,512 @@
+        generator_weight: weight of the generator's gan loss
+        discriminator_weight: weight of the discriminator gan loss
+        reload_path: path to a saved FullModelReplacement model to reload
+    """
+
+    generator_optimizer: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    discriminator_optimizer: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    generator: "RecurrentGeneratorConfig" = dataclasses.field(
+        default_factory=lambda: RecurrentGeneratorConfig()
+    )
+    discriminator: "DiscriminatorConfig" = dataclasses.field(
+        default_factory=lambda: DiscriminatorConfig()
+    )
+    convolution_type: str = "conv2d"
+    identity_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    target_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    gan_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    identity_weight: float = 1.0
+    target_weight: float = 1.0
+    generator_weight: float = 1.0
+    discriminator_weight: float = 1.0
+    reload_path: Optional[str] = None
+
+    def build(
+        self,
+        n_state: int,
+        nx: int,
+        ny: int,
+        n_batch: int,
+        n_time: int,
+        state_variables: Sequence[str],
+        scalers: Mapping[str, StandardScaler],
+    ) -> "FMRTrainer":
+        if self.reload_path is not None:
+            fmr_reload = FullModelReplacement.load(self.reload_path)
+            scalers = fmr_reload.scalers
+        if self.convolution_type == "conv2d":
+            convolution = single_tile_convolution
+        elif self.convolution_type == "halo_conv2d":
+            convolution = halo_convolution
+        else:
+            raise ValueError(f"convolution_type {self.convolution_type} not supported")
+        generator = self.generator.build(
+            n_state, nx=nx, ny=ny, n_time=n_time, convolution=convolution
+        ).to(DEVICE)
+        discriminator = self.discriminator.build(n_state, convolution=convolution,).to(
+            DEVICE
+        )
+        optimizer_generator = self.generator_optimizer.instance(generator.parameters())
+        optimizer_discriminator = self.discriminator_optimizer.instance(
+            discriminator.parameters()
+        )
+        init_weights(generator)
+        init_weights(discriminator)
+        fmr = FullModelReplacement(
+            model=FMRModule(generator=generator, discriminator=discriminator),
+            scalers=scalers,
+            state_variables=state_variables,
+        )
+        if self.reload_path is not None:
+            fmr.model.load_state_dict(fmr_reload.model.state_dict(), strict=True)
+        return FMRTrainer(
+            fmr=fmr,
+            optimizer_generator=optimizer_generator,
+            optimizer_discriminator=optimizer_discriminator,
+            identity_loss=self.identity_loss.instance,
```

Comment:
It should be fairly standard right now - `.instance` is used as a property when there are no arguments (normally because these are set by yaml config), and `.build(...)` is used when there are arguments (e.g. when shapes are determined from data automatically at runtime, and should not be specified in yaml config). I used something different for the model builder factories, but it's at least consistent across config classes.

Reply:
Thanks for clarifying the usages